src/core/ext/transport/chaotic_good/server/chaotic_good_server.cc

```diff
@@ -0,0 +1,370 @@
+  std::uniform_int_distribution<> distribution(0, charset.size() - 1);
+  random_string.reserve(8);
+  for (size_t i = 0; i < random_length; ++i) {
+    random_string += charset[distribution(generator)];
+  }
+  connection_id_ = Slice::FromCopiedString(random_string);
+}
+
+ChaoticGoodServerListener::ActiveConnection::HandshakingState::HandshakingState(
+    std::shared_ptr<ActiveConnection> connection)
+    : connection_(connection),
+      handshake_mgr_(std::make_shared<HandshakeManager>()),
+      arena_(connection_->arena_.get()),
+      context_(arena_.get()) {}
+
+void ChaoticGoodServerListener::ActiveConnection::HandshakingState::Start(
+    std::unique_ptr<EventEngine::Endpoint> endpoint) {
+  auto context = GetContext<Arena>();
+  GPR_ASSERT(context != nullptr);
+  handshake_mgr_->DoHandshake(
+      grpc_event_engine_endpoint_create(std::move(endpoint)),
+      connection_->args(), GetConnectionDeadline(connection_->args()), nullptr,
+      OnHandshakeDone, this);
+}
+void ChaoticGoodServerListener::ActiveConnection::HandshakingState::
+    OnHandshakeDone(void* arg, grpc_error_handle error) {
+  if (!error.ok()) {
+    gpr_log(GPR_ERROR, "Error server handshake failed: %s",
+            StatusToString(error).c_str());
+    return;
+  }
+  auto* args = static_cast<HandshakerArgs*>(arg);
+  GPR_ASSERT(args != nullptr);
+  if (args->endpoint == nullptr) {
+    gpr_log(GPR_ERROR, "Error server handshake done with null endpoint.");
+    return;
+  }
+  GPR_ASSERT(grpc_event_engine::experimental::grpc_is_event_engine_endpoint(
+      args->endpoint));
+  HandshakingState* self = static_cast<HandshakingState*>(args->user_data);
+  self->connection_->endpoint_ = std::make_shared<PromiseEndpoint>(
+      grpc_event_engine::experimental::grpc_take_wrapped_event_engine_endpoint(
+          args->endpoint),
+      SliceBuffer());
+  // Start receiving setting frames.
+  self->connection_->receive_settings_activity_ = OnReceive(self);
+}
+
+ActivityPtr
+ChaoticGoodServerListener::ActiveConnection::HandshakingState::OnReceive(
+    HandshakingState* self) {
+  auto on_receive_frames = TrySeq(
+      self->connection_->endpoint_->ReadSlice(FrameHeader::frame_header_size_),
+      [self = self](Slice slice) mutable {
+        // Parse frame header
+        auto frame_header = std::make_shared<FrameHeader>(
+            FrameHeader::Parse(reinterpret_cast<const uint8_t*>(
+                                   GRPC_SLICE_START_PTR(slice.c_slice())))
+                .value());
+        return TrySeq(
+            self->connection_->endpoint_->Read(frame_header->GetFrameLength()),
+            [frame_header = frame_header,
+             self = self](SliceBuffer buffer) mutable {
+              // Read Setting frame.
+              SettingsFrame frame;
+              // Initialized to get this_cpu() info in global_stat().
+              ExecCtx exec_ctx;
+              // Deserialize frame from read buffer.
+              absl::BitGen bitgen;
```

Comment:
use a member variable instead: these are expensive to initialize

Reply:
Done